### PR TITLE
Fixbuilderair

### LINKED ISSUE #457: Builder AI Freezing
1.11 beta v10 Minecolonies version

version minecraft 1.11.2

### Expected behavior

- Create Builder's Hut or Townhall (Any building really)

### Actual behaviour

- Builder stops building when running out of a material (Blocks, torches, etc.) part way through and just stands in one spot. (even when there are more blocks in builder's hut inventory) I'm able to solve the problem by handing blocks to the builder directly. No text prompt was given in chat.

### PR DESCRIPTION
**Pull request by xavierh**
Closes #457 

# Changes proposed in this pull request:
Sometimes the builder stop building waiting for an item he have in is chest and not on him.

This is because the builder have some AIR item in his inventory, while removing AIR item from the items needed, it will remove the first item, regardless to what it is,

This pr check for AIR item before removing it

The real questions are:
- Why *AIR Item*.isItemEqual(*any item*) is always true ?
- Why does the builder have air item in his inventory ?

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java [hunk 1]
```diff
@@ -288,6 +289,14 @@ public void addItemNeeded(@NotNull final ItemStack stack)
     public ItemStack removeItemNeeded(@NotNull final ItemStack stack)
     {
         @NotNull final ItemStack stackCopy = stack.copy();
+        //if stack is AIR, stack.isItemEqual(neededItem) will be always true
+        //and one itemNeeded would be remove
+        //There is probably a better way to check for air item
+        if (stack.getItemDamage()==0 && Item.getIdFromItem(stack.getItem()) == 0)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java]: Use `stack.isEmpty()` for this

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java (hunk 1 above)
Comment (xavierh) [reply src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java]: Ok will try this and report back

**Comment by forReason** [comment]
Wow that was quite a struggle I guess. How did you find out?

**Comment by marvin-bitterlich** [comment]
I wonder too, good find btw

**Review by xavierh — COMMENTED** [review]
It does not work with isEmpty. The stack is not empty it does have air in it

File: src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java]: did you try with stack == Itemstack.EmptyStack
?

**Comment by xavierh** [comment]
No, I will try.

I found by printing the list of needed items and the inventory in the sync
method. I noticed that when the problem occurred, the builder always had
air in his inventory.

Still does not explain why he's got air in the first place

On 29 Jan 2017 12:28, "Raycoms" <notifications@github.com> wrote:

*@Raycoms* commented on this pull request.
------------------------------

In src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java
<https://github.com/Minecolonies/minecolonies/pull/517>:

> @@ -288,6 +289,14 @@ public void addItemNeeded(@NotNull final ItemStack stack)
     public ItemStack removeItemNeeded(@NotNull final ItemStack stack)
     {
         @NotNull final ItemStack stackCopy = stack.copy();
+        //if stack is AIR, stack.isItemEqual(neededItem) will be always true
+        //and one itemNeeded would be remove
+        //There is probably a better way to check for air item
+        if (stack.getItemDamage()==0 &&
Item.getIdFromItem(stack.getItem()) == 0)

did you try with stack == Itemstack.EmptyStack
?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Minecolonies/minecolonies/pull/517>, or mute the thread

**Comment by Raycoms** [comment]
In 1.10 the inventory contained null stacks, in 1.11 they now contain air
stacks.

2017-01-29 11:25 GMT-02:00 xavierh <notifications@github.com>:

> No, I will try.
>
> I found by printing the list of needed items and the inventory in the sync
> method. I noticed that when the problem occurred, the builder always had
> air in his inventory.
>
> Still does not explain why he's got air in the first place
>

**Comment by xavierh** [comment]
stack == ItemStack.EMPTY does not work neither as I suspected

**Comment by Raycoms** [comment]
and check for stack == ItemStack.EMPTY || stack.getCount() == 0?

2017-01-29 14:14 GMT-02:00 xavierh <notifications@github.com>:

**Comment by marvin-bitterlich** [comment]
stack.isEmpty checks for null and == Stack.EMPTY, for Air, for <= 0 amount and for weird dmg values, so yeah that will work. If that does not work then there is something else wrong. I looked at the mc sources:

```
    public boolean isEmpty()
    {
        return this == EMPTY ? true : (this.item != null && this.item != Item.getItemFromBlock(Blocks.AIR) ? (this.stackSize <= 0 ? true : this.itemDamage < -32768 || this.itemDamage > 65535) : true);
    }
```

**Comment by xavierh** [comment]
Noted, will test that

**Comment by xavierh** [comment]
> stack == ItemStack.EMPTY || stack.getCount() == 0

seems to work

**Comment by marvin-bitterlich** [comment]
well that would stillt work with isEmpty xD

**Comment by xavierh** [comment]
Yes you are right isEmpty should work. I did !stack.isEmpty by mistake :)

**Comment by xavierh** [comment]
Ok, ready for review now

**Review by marvin-bitterlich — APPROVED** [review]

**Pull request merged**